lib/tasks/db/update_annotation_line.rake

```diff
@@ -0,0 +1,172 @@
+      SQL
+
+      end_time = Time.now
+      puts "1st sql for #{annotation_id__file_id__line_num.rows.length} annotations = #{end_time - start_time} seconds"
+      start_time = Time.now
+
+      file_annotation_lines = Hash.new { |hash, key| hash[key] = [] }
+      file_annotation_ids = Hash.new { |hash, key| hash[key] = [] }
+
+      annotation_id__file_id__line_num.rows.each do |row|
+        annotation_id, file_id, line_num = row
+        file_annotation_lines[file_id] << line_num
+        file_annotation_ids[file_id] << annotation_id
+      end
+
+      # Extract distinct file_id values from the file_annotation_lines hash
+      file_ids = file_annotation_lines.keys
+
+      end_time = Time.now
+      puts "extract file info = #{end_time - start_time} seconds"
+      start_time = Time.now
+
+      files = conn.exec_query(<<-SQL)
+        SELECT files.id, files.content, programming_questions.language_id
+        FROM course_assessment_answer_programming_files AS files
+        JOIN course_assessment_answers AS answers
+          ON files.answer_id = answers.actable_id
+          AND answers.actable_type = 'Course::Assessment::Answer::Programming'
+        JOIN course_assessment_questions AS questions
+          ON answers.question_id = questions.id
+        JOIN course_assessment_question_programming as programming_questions
+          ON questions.actable_id = programming_questions.id
+        WHERE files.id IN (#{file_ids.join(', ')})
+      SQL
+
+      end_time = Time.now
+      puts "2nd sql for #{file_ids.length} file contents = #{end_time - start_time} seconds"
+
+      start_time = Time.now
+
+      files_hash = files.rows.map { |row| { id: row[0], content: row[1], language_id: row[2] } }
+
+      language_ids = files_hash.map { |file| file[:language_id] }.uniq
+      languages = Coursemology::Polyglot::Language.where(id: language_ids).index_by(&:id)
+
+      annotation_ids_to_update = []
+
+      files_hash.each do |file|
+        content = file[:content]
+        language = languages[file[:language_id]]
+
+        highlighted_code = helper.highlight_code_block(content, language)
+
+        # skip to next unless new lines were inserted by application
+        next unless lines_are_added(highlighted_code)
+
+        annotation_lines = file_annotation_lines[file[:id]]
+        smallest_decrementable_number = smallest_decrementable_line_number(annotation_lines)
+
+        # if all the annotations in the file is consecutive from line 1, skip this file
+        next if smallest_decrementable_number == annotation_lines.length + 1
+
+        annotation_ids = file_annotation_ids[file[:id]]
+        annotation_line_to_annotation_id = annotation_lines.zip(annotation_ids).to_h
+        annotation_lines.each do |line|
+          # only decrement line that is after the smallest decrementable number
+          # and also if there is no annotation in the line directly preceding that
+          next if line <= smallest_decrementable_number
```

Comment:
why should I next if `line == smallest_decremental_number`?

Reply:
`smallest_decremental_number` is actually the smallest number in which the annotation line is not present (so in case the annotation line is `[1, 2, 3, 7, 8, 9]`, the number will be `4`..

so, actually with or without the equal sign, it won't matter as the case where `line == smallest_decrementable_number` won't exist